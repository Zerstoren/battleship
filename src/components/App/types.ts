import { ILobbyStore, IMainStore } from "../../stores/types";

export enum GameStatus {
  MAIN = 'main',
  LOBBY_LIST = 'list',
  CREATE_LOBBY = 'create',
  WAIT_CONNECT = 'wait',
  SET_SHIPS = 'set_ships',
  GAME = 'game_process',
  GAMEOVER = 'game_end',
}

export interface IGameState {
  gameStatus: GameStatus,
  gameWithUser: boolean,

  setGameStatus: (status: GameStatus) => void,
}

export interface IAppProps {
  mainStore?: IMainStore
}

export interface IProps {
  handleChangeGameStatus: (startGame: GameStatus) => void,
}

export interface ISetLobbyProps {
  handleSetLobby: (startGame: GameStatus, lobby: ILobbyStore) => void
}