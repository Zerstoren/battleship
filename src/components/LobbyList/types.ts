import { ISnapshotOutLobbyStore } from "../../stores/types";

export interface IDataFromServer {
  [key: string]: ISnapshotOutLobbyStore
}
