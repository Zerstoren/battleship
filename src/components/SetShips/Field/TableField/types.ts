import { DragObjectWithType } from "react-dnd/lib/interfaces/hooksApi";

export enum MatrixFill {
  EMPTY = '',
  SHADOW = 's',
  ERR_SHADOW = 'e',
  SET = 'w',
  SET_KILL = 'k',
  MISS = 'm',
}

export interface IMatrix extends Array<Array<MatrixFill>> {}

export interface DragObjectItem extends DragObjectWithType {
  size: number
}