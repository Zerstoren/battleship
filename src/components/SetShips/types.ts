import { IMainStore } from "../../stores/mainStore";
import { IProps as IAppProps } from "../App/types";

export interface IProps extends IAppProps {
  mainStore?: IMainStore
}