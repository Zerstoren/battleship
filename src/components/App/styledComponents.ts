import styled from "styled-components";

export const CenteredDiv = styled.div`
  width: 600px;
  margin: 0 auto;
  text-align: center;
`;