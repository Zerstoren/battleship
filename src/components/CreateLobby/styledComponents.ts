import styled from "styled-components";

export const BlockSpacing = styled.div`
padding: 10px 0;
`;

export const FormFloating = styled.div.attrs(props => ({
  className: "form-floating"
}))``;

export const FloatingInput = styled.input.attrs(props => ({
  className: `form-control ${props.className}`
}))``;