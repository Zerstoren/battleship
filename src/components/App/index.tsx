import { inject, observer } from 'mobx-react';
import React, { FC } from 'react';
import { ILobbyStore } from '../../stores/lobby';
import CreateLobby from '../CreateLobby';
import LobbyList from '../LobbyList';
import Main from '../MainPage';
import SetShips from '../SetShips';
import WaitConnect from '../WaitConnect';
import { AlertError, CenteredDiv } from './styledComponents';
import { GameStatus, IAppProps } from './types';

const App: FC<IAppProps> = inject("mainStore")(observer((props) => {
  const state = props.mainStore;

  const handleGameStart = (startGame: GameStatus) => state!.setGameStatus(startGame);
  const handleSetLobby = (startGame: GameStatus, lobby: ILobbyStore) => state!.setLobby(startGame, lobby);
  const handleWaitLobby = (startGame: GameStatus, lobby: ILobbyStore | null = null) => state!.setLobby(startGame, lobby);
  const handleAcceptError = () => state?.setError('');

  let component = null;
  
  switch(state!.status) {
    case GameStatus.MAIN:
      component = (<Main handleChangeGameStatus={handleGameStart} />);
      break;
    
    case GameStatus.CREATE_LOBBY:
      component = (<CreateLobby handleChangeGameStatus={handleWaitLobby} />);
      break;

    case GameStatus.WAIT_CONNECT:
      component = (<WaitConnect handleChangeGameStatus={handleGameStart} />);
      break;

    case GameStatus.LOBBY_LIST:
      component = (<LobbyList handleSetLobby={handleSetLobby} />);
      break;

    case GameStatus.SET_SHIPS:
      component = (<SetShips handleChangeGameStatus={handleGameStart} />);
  }

  let err = null;
  
  if (state!.error) {
    err = <AlertError>
      <strong>{state!.error}</strong>
      <button type="button" className="btn-close" onClick={handleAcceptError}></button>
    </AlertError>;
  }

  return (
    <>
      {err}
      <CenteredDiv>
        {component}
      </CenteredDiv>
    </>
  );
}));

export default App;
