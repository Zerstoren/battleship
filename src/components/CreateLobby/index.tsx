import { getSnapshot } from 'mobx-state-tree';
import React, { FC } from 'react';
import { Field, Form } from 'react-final-form';
import { AppHeader } from '../../shared/StyledComponents/Headers';
import lobbyStore from '../../stores/lobby';
import { ISnapshotOutLobbyStore } from '../../stores/types';
import { GameStatus, IProps } from '../App/types';
import { onValidate } from './helpersFns';
import { BlockSpacing, FloatingInput, FormFloating } from './styledComponents';

const CreateLobby: FC<IProps> = (props) => {
  const {handleChangeGameStatus} = props;
  const store = lobbyStore.create({
    lobbyName: ''
  });

  const backToMainPage = () => {
    handleChangeGameStatus(GameStatus.MAIN);
  };

  const handleSubmit = (values: ISnapshotOutLobbyStore) => {
    const store = lobbyStore.create(values);
    store.publishLobby();
    handleChangeGameStatus(GameStatus.WAIT_CONNECT);
  };

  return (
    <Form 
      initialValues={getSnapshot(store)}
      onSubmit={handleSubmit}
      validate={onValidate}
      render={(props => (
        <>
          <AppHeader>New Lobby</AppHeader>

          <BlockSpacing>
            <Field name="lobbyName">
              {({input, meta}) => (
                <FormFloating>
                  <FloatingInput type="text" autoFocus {...input} className={meta.error && meta.touched && 'is-invalid'} />
                  <label>{(meta.error && meta.touched) ? meta.error : "Lobby Name"}</label>
                </FormFloating>
              )}
            </Field>
          </BlockSpacing>

          <BlockSpacing className="d-flex justify-content-between">
            <Field name="x">
              {({input, meta}) => (
                <FormFloating>
                  <FloatingInput type="number" className={meta.error && meta.touched && 'is-invalid'} {...input} />
                  <label>{(meta.error && meta.touched) ? meta.error : "Field size: X"}</label>
                </FormFloating>
              )}
            </Field>
            <Field name="y">
              {({input, meta}) => (
                <FormFloating>
                  <FloatingInput type="number" className={meta.error && meta.touched && 'is-invalid'} {...input} />
                  <label>{(meta.error && meta.touched) ? meta.error : "Field size: Y"}</label>
                </FormFloating>
              )}
            </Field>
          </BlockSpacing>
          
          <BlockSpacing className="d-flex justify-content-between">
            <Field name="ships4n">
              {({input, meta}) => (
                <FormFloating>
                  <FloatingInput className={meta.error && 'is-invalid'} type="number" {...input} />
                  <label>{(meta.error) ? meta.error : "4x ships count"}</label>
                </FormFloating>
              )}
            </Field>

            <Field name="ships3n">
              {({input, meta}) => (
                <FormFloating>
                  <FloatingInput className={meta.error && 'is-invalid'} type="number" {...input} />
                  <label>{(meta.error) ? meta.error : "3x ships count"}</label>
                </FormFloating>
              )}
            </Field>
          </BlockSpacing>
          
          <BlockSpacing className="d-flex justify-content-between">
            <Field name="ships2n">
              {({input, meta}) => (
                <FormFloating>
                  <FloatingInput className={meta.error && 'is-invalid'} type="number" {...input} />
                  <label>{(meta.error) ? meta.error : "2x ships count"}</label>
                </FormFloating>
              )}
            </Field>
            
            <Field name="ships1n">
              {({input, meta}) => (
                <FormFloating>
                  <FloatingInput className={meta.error && 'is-invalid'} type="number" {...input} />
                  <label>{(meta.error) ? meta.error : "1x ships count"}</label>
                </FormFloating>
              )}
            </Field>
          </BlockSpacing>
          
          <BlockSpacing className="d-flex justify-content-evenly">
            <button className="btn btn-secondary" onClick={backToMainPage}>Back</button>
            <button className="btn btn-primary" onClick={props.handleSubmit}>Create lobby</button>
          </BlockSpacing>
        </>
      ))}
    />
  );
};

export default CreateLobby;