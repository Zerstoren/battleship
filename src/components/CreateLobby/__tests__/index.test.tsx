import React from 'react';
import {mount} from 'enzyme';
import CreateLobby from '..';
import { GameStatus } from '../../App/types';

describe('CreateLobby', () => {
  test('Type all data', () => {
    const mockCallback = jest.fn((state, lobby) => null);

    const tree = mount(<CreateLobby handleChangeGameStatus={mockCallback} />);
    expect(tree.debug()).toMatchSnapshot();

    tree.find('input[name="lobbyName"]').simulate("change", { target: { value: "Whooo" }})
    tree.find('input[name="x"]').simulate("change", { target: { value: "20" }})
    tree.find('input[name="y"]').simulate("change", { target: { value: "21" }})
    tree.find('input[name="ships4n"]').simulate("change", { target: { value: "2" }})
    tree.find('input[name="ships3n"]').simulate("change", { target: { value: "3" }})
    tree.find('input[name="ships2n"]').simulate("change", { target: { value: "4" }})
    tree.find('input[name="ships1n"]').simulate("change", { target: { value: "5" }})
    tree.find('button.btn-primary').simulate('click');
    
    const lobby = mockCallback.mock.calls[0][1];
    expect(mockCallback.mock.calls[0][0]).toBe(GameStatus.WAIT_CONNECT);
    expect(lobby.lobbyName).toBe('Whooo');
    expect(lobby.x).toBe(20);
    expect(lobby.y).toBe(21);
    expect(lobby.ships4n).toBe(2);
    expect(lobby.ships3n).toBe(3);
    expect(lobby.ships2n).toBe(4);
    expect(lobby.ships1n).toBe(5);
  });

  test('Check validation name', () => {
    const mockCallback = jest.fn((state, lobby) => null);
    const tree = mount(<CreateLobby handleChangeGameStatus={mockCallback} />);

    tree.find('button.btn-primary').simulate('click');
    expect(mockCallback.mock.calls.length).toBe(0);
    expect(tree.find('input.is-invalid').length).toBe(1);
    expect(tree.debug()).toMatchSnapshot();
  });

  test('Check validation field size', () => {
    const mockCallback = jest.fn((state, lobby) => null);
    const tree = mount(<CreateLobby handleChangeGameStatus={mockCallback} />);

    tree.find('input[name="lobbyName"]').simulate("change", { target: { value: "Whooo" }})
    tree.find('input[name="x"]').simulate("change", { target: { value: "1" }})
    tree.find('input[name="y"]').simulate("change", { target: { value: "1" }})
    tree.find('button.btn-primary').simulate('click');
    expect(mockCallback.mock.calls.length).toBe(0);
    expect(tree.find('input.is-invalid').length).toBe(6);
    expect(tree.debug()).toMatchSnapshot();
  });

  test('Check validation ships size', () => {
    const mockCallback = jest.fn((state, lobby) => null);
    const tree = mount(<CreateLobby handleChangeGameStatus={mockCallback} />);

    tree.find('input[name="lobbyName"]').simulate("change", { target: { value: "Whooo" }})
    tree.find('input[name="ships4n"]').simulate("change", { target: { value: "22" }})
    tree.find('button.btn-primary').simulate('click');
    expect(mockCallback.mock.calls.length).toBe(0);
    expect(tree.find('input.is-invalid').length).toBe(4);
    expect(tree.debug()).toMatchSnapshot();
  });
});

