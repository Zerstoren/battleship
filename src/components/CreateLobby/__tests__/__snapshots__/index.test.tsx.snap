// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateLobby Check validation field size 1`] = `
"<h1 class=\\"sc-bdfBwQ display-3\\">New Lobby</h1>
<div class=\\"sc-gsTCUz hIfhYJ\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input name=\\"lobbyName\\" class=\\"sc-hKgILt form-control undefined\\" type=\\"text\\" value=\\"Whooo\\"><label>Lobby Name</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt is-invalid form-control is-invalid\\" name=\\"x\\" value=\\"1\\"><label>Field is too short</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt is-invalid form-control is-invalid\\" name=\\"y\\" value=\\"1\\"><label>Field is too short</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships4n\\" value=\\"1\\"><label>Need less ships</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships3n\\" value=\\"2\\"><label>Need less ships</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships2n\\" value=\\"3\\"><label>Need less ships</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships1n\\" value=\\"4\\"><label>Need less ships</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-evenly\\"><button class=\\"btn btn-secondary\\">Back</button><button class=\\"btn btn-primary\\">Create lobby</button></div>"
`;

exports[`CreateLobby Check validation name 1`] = `
"<h1 class=\\"sc-bdfBwQ display-3\\">New Lobby</h1>
<div class=\\"sc-gsTCUz hIfhYJ\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input name=\\"lobbyName\\" class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"text\\" value=\\"\\"><label>Lobby name is required</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt form-control undefined\\" name=\\"x\\" value=\\"10\\"><label>Field size: X</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt form-control undefined\\" name=\\"y\\" value=\\"10\\"><label>Field size: Y</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships4n\\" value=\\"1\\"><label>4x ships count</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships3n\\" value=\\"2\\"><label>3x ships count</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships2n\\" value=\\"3\\"><label>2x ships count</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships1n\\" value=\\"4\\"><label>1x ships count</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-evenly\\"><button class=\\"btn btn-secondary\\">Back</button><button class=\\"btn btn-primary\\">Create lobby</button></div>"
`;

exports[`CreateLobby Check validation ships size 1`] = `
"<h1 class=\\"sc-bdfBwQ display-3\\">New Lobby</h1>
<div class=\\"sc-gsTCUz hIfhYJ\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input name=\\"lobbyName\\" class=\\"sc-hKgILt form-control undefined\\" type=\\"text\\" value=\\"Whooo\\"><label>Lobby Name</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt form-control undefined\\" name=\\"x\\" value=\\"10\\"><label>Field size: X</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt form-control undefined\\" name=\\"y\\" value=\\"10\\"><label>Field size: Y</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships4n\\" value=\\"22\\"><label>Need less ships</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships3n\\" value=\\"2\\"><label>Need less ships</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships2n\\" value=\\"3\\"><label>Need less ships</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt is-invalid form-control is-invalid\\" type=\\"number\\" name=\\"ships1n\\" value=\\"4\\"><label>Need less ships</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-evenly\\"><button class=\\"btn btn-secondary\\">Back</button><button class=\\"btn btn-primary\\">Create lobby</button></div>"
`;

exports[`CreateLobby Type all data 1`] = `
"<h1 class=\\"sc-bdfBwQ display-3\\">New Lobby</h1>
<div class=\\"sc-gsTCUz hIfhYJ\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input name=\\"lobbyName\\" class=\\"sc-hKgILt form-control false\\" type=\\"text\\" value=\\"\\"><label>Lobby Name</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt form-control undefined\\" name=\\"x\\" value=\\"10\\"><label>Field size: X</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input type=\\"number\\" class=\\"sc-hKgILt form-control undefined\\" name=\\"y\\" value=\\"10\\"><label>Field size: Y</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships4n\\" value=\\"1\\"><label>4x ships count</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships3n\\" value=\\"2\\"><label>3x ships count</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-between\\">
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships2n\\" value=\\"3\\"><label>2x ships count</label></div>
    <div class=\\"sc-dlfnbm form-floating\\"><input class=\\"sc-hKgILt form-control undefined\\" type=\\"number\\" name=\\"ships1n\\" value=\\"4\\"><label>1x ships count</label></div>
</div>
<div class=\\"sc-gsTCUz hIfhYJ d-flex justify-content-evenly\\"><button class=\\"btn btn-secondary\\">Back</button><button class=\\"btn btn-primary\\">Create lobby</button></div>"
`;
